src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs

```diff
@@ -0,0 +1,401 @@
+                return;
+
+            var cpsUpdateService = await GetUpdateServiceIfCpsProjectAsync(project, cancellationToken).ConfigureAwait(false);
+            var task = cpsUpdateService != null
+                ? NotifyCpsProjectSystemAsync(project, cpsUpdateService, attributeInfos, cancellationToken)
+                : NotifyLegacyProjectSystemAsync(project, attributeInfos, cancellationToken);
+
+            await task.ConfigureAwait(false);
+        }
+
+        private Task NotifyLegacyProjectSystemAsync(
+            Project project,
+            IEnumerable<DesignerInfo> attributeInfos,
+            CancellationToken cancellationToken)
+        {
+            // Move over to the UI thread and attempt to notify it about all the documents from this
+            // particular project at once. Use a task completion source here so we can kick over to
+            // the UI thread, but still ensure that our caller knows when we complete.
+            var completionSource = new TaskCompletionSource<bool>();
+
+            _notificationService.RegisterNotification(() =>
+            {
+                AssertIsForeground();
+
+                try
+                {
+                    NotifyLegacyProjectSystemOnUIThread(project, attributeInfos, cancellationToken);
+                }
+                catch (OperationCanceledException)
+                {
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+                {
+                }
+                finally
+                {
+                    completionSource.SetResult(true);
+                }
+            }, _listener.BeginAsyncOperation("RegisterDesignerAttribute"));
+
+            return completionSource.Task;
+        }
+
+        private void NotifyLegacyProjectSystemOnUIThread(
+            Project project,
+            IEnumerable<DesignerInfo> attributeInfos,
+            CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            var designerService = GetDesignerServiceOnForegroundThread();
+            if (designerService == null)
+                return;
+
+            var hierarchy = _workspace.GetHierarchy(project.Id);
+            if (hierarchy == null)
+                return;
+
+            foreach (var info in attributeInfos)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                NotifyLegacyProjectSystemOnUIThread(
+                    project, designerService, hierarchy, info);
+            }
+        }
+
+        private void NotifyLegacyProjectSystemOnUIThread(
+            Project project,
+            IVSMDDesignerService designerService,
+            IVsHierarchy hierarchy,
+            DesignerInfo info)
+        {
+            // Make sure this is a document we still know about when OOP notifies us.
+            var document = project.GetDocument(info.DocumentId);
+            if (document == null)
+                return;
```

Comment:
the rest of the code in this method matches what we previously did in the last incremental analyzer.  the code is copied almost unchanged.